flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyGroupedPriorityQueue.java

```diff
@@ -18,24 +18,36 @@
 
 package org.apache.flink.state.changelog;
 
+import org.apache.flink.api.common.typeutils.TypeSerializer;
 import org.apache.flink.runtime.state.KeyGroupedInternalPriorityQueue;
 import org.apache.flink.util.CloseableIterator;
+import org.apache.flink.util.ExceptionUtils;
 
 import javax.annotation.Nullable;
 
+import java.io.IOException;
 import java.util.Collection;
 import java.util.Set;
 
+import static org.apache.flink.state.changelog.StateChangeLogger.loggingIterator;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
 /**
  * A {@link KeyGroupedInternalPriorityQueue} that keeps state on the underlying delegated {@link
  * KeyGroupedInternalPriorityQueue} as well as on the state change log.
  */
 public class ChangelogKeyGroupedPriorityQueue<T> implements KeyGroupedInternalPriorityQueue<T> {
     private final KeyGroupedInternalPriorityQueue<T> delegatedPriorityQueue;
+    private final PqStateChangeLogger<T> logger;
```

Comment:
Are these updates writing to the same log? I mean they are using different writers, but they might share the same log write?

If the scope is not specified, how we can differentiate this PQ from the normal change log?

I understand with normal changes can be differentiated using key context, but how is this one?

Reply:
State type is written as part of the metadata (and for each state change, state name is written, so we can relate it to the metadata).

The current key group is written in `AbstractStateChangeLogger.log`, so we can differentiate between the groups on rescaling.